### PR TITLE
Feature/kmeansandPAM

### PR DESCRIPTION
**Pull request by NickHager**
K-Means and PAM added, changes to driver to run tests, as well.

### CODE REVIEW

File: Project 2/k-NNAlgorithms/src/AbaloneData.java [hunk 1]
```diff
@@ -16,5 +16,23 @@ private void preProcess(){
                 in a meaningful way */
             fullSet.get(i).remove(0);
         }
+
+        for (int i = 0; i <fullSet.size() ; i++) {
+            Double classification = Double.parseDouble(fullSet.get(i).get(fullSet.get(0).size()-1));
+            //if num rings less than 10, set to class 0, young
+            if(classification<8){
```
Comment (RialJohnson) [inline Project 2/k-NNAlgorithms/src/AbaloneData.java]: 8 is a magic number

File: Project 2/k-NNAlgorithms/src/AbaloneData.java [hunk 2]
```diff
@@ -16,5 +16,23 @@ private void preProcess(){
...
+                fullSet.get(i).set(fullSet.get(0).size()-1,"0");
+            }
+            //if num rings greater than or equal to 10 but less than 20, set to class 1, intermediate
+            else if(classification<12){
```
Comment (RialJohnson) [inline Project 2/k-NNAlgorithms/src/AbaloneData.java]: 12 is a magic number

File: Project 2/k-NNAlgorithms/src/Algorithms.java [hunk 3]
```diff
@@ -183,6 +190,161 @@
...
+    public static ArrayList<ArrayList<String>> Kmeans (ArrayList<ArrayList<String>>trainingData, int numClasses){
+        // Variables
+        ArrayList<ArrayList<String>> clusterCentroids = new ArrayList<>();
+        int numFeatures= trainingData.get(0).size() - 1;
+        // 1) initialize cluster centroids randomly
+        for (int i = 0; i < numClasses; i++){ // Number of cluster centroids = number of classes
+            clusterCentroids.add(MathFunction.randomCentroid(numFeatures));
+        }
+        int preventEndless = 0; // Remove for final code
+        while (preventEndless < 50000) { // convert to while (true)
```
Comment (RialJohnson) [inline Project 2/k-NNAlgorithms/src/Algorithms.java:205]: what is 50000?

**Review by Rusty89 — APPROVED** [review]
Looks good to me

**Review by RialJohnson — APPROVED** [review]

**Comment by GiantFrog** [comment]
this is censorship!

**Pull request merged**